File: website/docs/tutorials/authentication-and-security/authentication.py
# skip:start
from typegraph import typegraph, Policy, t, Graph
from typegraph.graph.params import Cors, Auth
from typegraph.runtimes.deno import DenoRuntime

# skip:end


@typegraph(
    auths=[
        # highlight-start
        # expects a secret in metatype.yml
        # `TG_[typegraph]_BASIC_[username]`
        # highlight-next-line
        Auth.basic(["admin"]),
        # highlight-end
    ],
    # skip:next-line
    cors=Cors(allow_origin=["https://metatype.dev", "http://localhost:3000"]),
)
def authentication(g: Graph):
    deno = DenoRuntime()
    public = Policy.public()

    ctx = t.struct({"username": t.string().optional().from_context("username")})

    g.expose(
        get_context=deno.identity(ctx),
        default_policy=[public],
    )

File: website/typegraphs/authentication.ts
// skip:start
import { Policy, t, typegraph } from "@typegraph/sdk/index.js";
import { Auth } from "@typegraph/sdk/params.js";
import { DenoRuntime } from "@typegraph/sdk/runtimes/deno.js";

// skip:end

typegraph({
  name: "authentication",
}, (g) => {
  const deno = new DenoRuntime();
  const pub = Policy.public();

  const ctx = t.struct({
    "username": t.string().optional().fromContext("username"),
  });

  // highlight-start
  // expects a secret in metatype.yml
  // `TG_[typegraph]_BASIC_[username]`
  // highlight-next-line
  g.auth(Auth.basic(["admin"]));
  // highlight-end

  g.expose({
    get_context: deno.identity(ctx).withPolicy(pub),
  });
});
